src/apps/parser.c

```diff
@@ -0,0 +1,193 @@
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** \file
+ * Parser demo app
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+#include <repgp/repgp.h>
+
+#include <rnp/rnp.h>             // for rnp_t, rnp_ctx_t et. all
+#include <rekey/rnp_key_store.h> // for keystore stuff
+
+#define RING "keyring"
+
+static bool
+configure_rnp(rnp_t *rnp)
+{
+    rnp_params_t params = {0};
+    bool         res = true;
+
+    // IO
+    pgp_io_t pgpio = {.errs = stderr, .res = stdout, .outs = stdout};
+    rnp->io = malloc(sizeof(pgp_io_t));
+    if (!rnp->io) {
+        return false;
+    }
+
+    memcpy(rnp->io, &pgpio, sizeof(pgp_io_t));
+
+    // RNP
+    params.pubpath = strdup(RING "pubring.gpg");
+    params.secpath = strdup(RING "secring.gpg");
+    if (!params.pubpath || !params.secpath) {
+        res = false;
+        goto end;
+    }
+    params.ks_pub_format = RNP_KEYSTORE_GPG;
+    params.ks_sec_format = RNP_KEYSTORE_GPG;
+    if (rnp_init(rnp, &params) != RNP_SUCCESS) {
+        res = false;
+        goto end;
+    }
+
+    // Load keys
+    if (!rnp_key_store_load_keys(rnp, true)) {
+        res = false;
+        goto end;
+    }
+
+end:
+    rnp_params_free(&params);
+    return res;
+}
+
+void
+verification()
+{
+    rnp_t     rnp = {0};
+    rnp_ctx_t ctx = {0};
+
+    assert(configure_rnp(&rnp));
+    rnp_ctx_init(&ctx, &rnp);
+
+    repgp_io_t io = repgp_create_io();
+    assert(io != REPGP_HANDLE_NULL);
+
+    repgp_set_input(io, create_filepath_handle("signed.gpg"));
+    repgp_set_output(io, create_filepath_handle("signed_out"));
+    printf("RES = %d\n", repgp_verify(&ctx, io) == RNP_SUCCESS);
+
+end:
+    repgp_destroy_io(io);
+    free(rnp.io);
+    rnp.io = NULL;
+    rnp_end(&rnp);
+}
+
+void
+decryption()
+{
+    rnp_t     rnp = {0};
+    rnp_ctx_t ctx = {0};
+
+    assert(configure_rnp(&rnp));
+    rnp_ctx_init(&ctx, &rnp);
+
+    /* ----- perform decryption ----- */
+    uint8_t out_buf[4096] = {0};
+    size_t  out_buf_size = sizeof(out_buf);
+    uint8_t in_buf[4096] = {0};
+    size_t  in_buf_size = sizeof(in_buf);
+
+    FILE *f = fopen("encrypted.gpg", "rb");
+    assert(f != NULL);
+    in_buf_size = fread(in_buf, 1, in_buf_size, f);
+    assert(in_buf_size > 0);
```

Comment:
I totally dissagree to use assert for runtime-error checking because assert should/may be dissabled by adding `-DNDEBUG` to `cc` flags and this checking may exist only on debug build.